### DISCUSSION TITLE
Error Handling for Commands

### DISCUSSION THREAD

**Discussion by mark-studer** [PowerShell/Crescendo#115] [Ideas]
What is the expectation around error handling for commands?  I often wrap native command so I can get good error handling for automation situations.  This generally means checking the $LASTEXITCODE to validate the command completed successfully.  I did a test wrapping git and if the command fails you get no error.  Looking at the generated command it has no error checking around the execution.  Is there a way to enable that in the .json file or a plan for adding built-in or opt-in error handling?

**Comment by theJasonHelmick** [comment]
You raise an excellent point and we should consider error handling. I haven't added this to a milestone yet, but let's open the discussion here and get everyones feedback regarding concerns and implementation ideas.

**Comment by ethanbergstrom** [comment]
+1 on error handling. Chocolatey packages, for example, allow the package owner to specify a list of successful exit codes, and everything else gets treated as an error. Without a programmatic way to catch and handle error signals returned by the native command, we're left trying to guess how to parse the native command output for "uniquely error-like" text and throw an exception in the handler.

**Comment by kstrauss** [comment]
A few things that occur to me:
1. need to be able to specify a list of "good" return codes or a list of bad ones. Some applications have a return code say that means finished with warnings. Not sure how to represent other than perhaps a flag of TreatWarningsLikeErrors
2. Need to be able to ignore the return code and instead delegate to a function/script that deduces it from output (because the return code is always 0). Ideally, I would like a set of established ones (i.e. any output to stderr is an error, regex for pattern)
3. potentially add a list of handling strategies (throw exception, have a retry, retry with backoff, timeout, etc. like exists in [Polly](https://github.com/App-vNext/Polly#resilience-policies) ) that should be added as parameters/switches on the generated wrapper. Should still support standard parameters like ErrorVariable

**Comment by JamesWTruher** [comment]
there's not much here that's really actionable, so I've turned this into a discussion.
While I agree that there may be something here that could be pretty cool, it needs a lot more clarity wrt what it should actually _do_. Also, since we allow pipelining to an output handler, and that output handler _may_ also include calls native executables it becomes so much more complicated.

The one approach I can see would be along the lines of the following:
- given a _non-streaming_ Crescendo function, provide a scriptblock/function/script per `$LASTEXITCODE` which will be called (the default case of 0 being just hand the result to the output handler)

The whole error delegate suggestion seems very fragile and over complicated, i'd love to know how this would work in practice.
 
off the top of my head i see something like:

```json
"errorhandling": [
  {
     "lastExitCode": 3,
     "handlerType": "inline",
     "handler": "throw 'whoops - badness has ensued'"
  },
]
```

when the LASTEXITCODE matches, the error handler is called before calling the output handler. The error handler can receive the output of the faulting command so it can do whatever it wants. If the error handler throws or emits a terminating error, the output handler is not called, otherwise whatever output is handed to the outputHandler.

It seems to me that retrying or whatever faulted situation becomes _ridiculously_ complicated to do generically.

**Reply by ethanbergstrom** [reply]
I agree that trying to capture retry logic or other options is outside of scope, and think that should be handled by the caller to the cmdlet if it's needed.

Instead of trying to capture and handle every possible error code, I would flip it around and give the output handler definition the ability to set an array of known 'good' exit codes (defaulting to 0 if not otherwise set). If the output handler code doesn't match, then fall back to the error handler, and maybe the error handler (again, by default) just Throws or Write-Errors whatever it gets unless someone feels they need to write their own special handler to capture and format the error output. 

The objective of the output handler is to parse strings into objects, but if something failed, I'm not going to be expecting an object back anyway, and would want to know why it failed, so having the full output get returned to the user as an error after a non-zero exit code by default I think would make sense and save everyone a lot of boilerplate work.

**Reply by ethanbergstrom** [reply]
Regarding streaming vs. non-streaming, and back to my above thought about having a default error handler (unless someone writes their own) - maybe by default Write-Error (non-terminating) the output if its streaming and Throw (terminating) the output if its non-streaming?

**Reply by mark-studer** [reply]
I would agree with @ethanbergstrom on how I would expect this to work.  Assuming you pass in the exit code to the error handler the developer can implement any switching logic based on different codes in their handler.  I think 99% of the time they only really care about was it a success or did it fail so having a single list of success response codes would make the common case very easy but allow complete flexibility if needed.

Not sure about the streaming how you would detect that as I assume you wouldn't have an exit code as the app hasn't exited... If there is a way to handle that great, but I think this could be focused on non-streaming.
